src/test/connect-test.c

```
@@ -0,0 +1,444 @@
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+#include <fcntl.h>
+
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+
+#include "sd-event.h"
+#include "time-util.h"
+#include "string-util.h"
+#include "parse-util.h"
+#include "socket-util.h"
+
+#define INPUT_CONNECTION_SUCCESS  "success"
+#define INPUT_CONNECTION_REJECTED "rejected"
+#define INPUT_CONNECTION_TIMEOUT  "timeout"
+
+enum connection_status {
+        CONNECTION_UNDEFINED = 0,
+        CONNECTION_TIMEOUT,
+        CONNECTION_REFUSED,
+        CONNECTION_SUCCEEDED,
+};
+
+struct context {
+        enum connection_status ipv4;
+        enum connection_status ipv6;
+        enum connection_status timeout;
+};
+
+static const char *enum_to_s(enum connection_status status) {
+        switch (status) {
+        case CONNECTION_TIMEOUT:
+                return INPUT_CONNECTION_TIMEOUT;
+        case CONNECTION_REFUSED:
+                return INPUT_CONNECTION_REJECTED;
+        case CONNECTION_SUCCEEDED:
+                return INPUT_CONNECTION_SUCCESS;
+        default:
+                return "undefined";
+        }
+}
+
+static int timeout_handler(sd_event_source *event_source, usec_t usec, void *data) {
+        struct context *ctx = data;
+
+        ctx->timeout = CONNECTION_TIMEOUT;
+
+        sd_event_exit(sd_event_source_get_event(event_source), 0);
+        return 0;
+}
+
+static int server_handler(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        int conn_fd, r;
+        union sockaddr_union addr = { 0 };
+        struct sockaddr *sa, addr_name = { 0 };
+        socklen_t addr_length = sizeof(addr), len;
+
+        r = getsockname(fd, &addr_name, &addr_length);
+        if (r < 0) {
+                sd_event_exit(sd_event_source_get_event(event_source), r);
+                return -1;
```

we usually try to avoid making up errors such as "-1". Please just propagate the actual error you are seeing here, hence do "return -errno" here.